creme/utils/shift.py

```diff
@@ -0,0 +1,100 @@
+import collections
+import typing
+
+__all__ = ['Shift']
+
+
+class DefaultStat:
+    """Default stat when using Shift class"""
+
+    def __init__(self):
+        self.x = None
+
+    def update(self, x):
+        self.x = x
+        return self
+
+    def get(self):
+        return self.x
+
+    @property
+    def name(self):
+        return ''
+
+
+class Shift:
+    """Shift the observations with the selected period.
+
+    Shift class allows to store and recall at regular intervals values observed in a stream.
+    The method ```get()`` returns the ```n-th - period``` observed entry.
+
+    Parameters:
+        period (int): Number of periods to shift. Must be positive.
+        stats (stats.Univariate, None): The statistic to compute. When set to None, shift the input
+            series of values without updating any statistics.
+        missing: Handle missing values when there is'nt any reference for a given period.
+        fitting_before_transform (bool): Shift logic. Must be set to True when the user updates the
+            buffer before retrieving the observation with the offset selected.
+
+    Example:
+
+        ::
+
+            # Example 1:
+            >>> from creme import utils
+
+            >>> shift = utils.Shift(period=2, missing=-1)
+
+            >>> X = [
+            ...    1,
+            ...    2,
+            ...    3,
+            ...    4,
+            ...    5,
+            ... ]
+
+            >>> for x in X:
+            ...    print(shift.get())
+            ...    _ = shift.update(x)
+            -1
+            -1
+            1
+            2
+            3
+
+            >>> from creme import stats
+            >>> shift = utils.Shift(period=2, stats=stats.Mean(), missing=-1.)
+
+            >>> X = [
+            ...    1,
+            ...    2,
+            ...    3,
+            ...    4,
+            ...    5,
+            ... ]
+
+    """
+    def __init__(self, period: int, stats=DefaultStat(), missing: float=0.,
```

Comment:
Is there any point of having `DefaultStat` as a default?

Also now that I see the code I think we can do something better. As far as I can tell is that we want to compose `Shift` and `Mean`. Therefore, instead of passing a `stat` to `Shift`, we could instead implement a pipe operator to be able to chain statistics. Therefore we could write something like `Shift(1) | Mean()`. That would be cool because it means we don't have to modify the code of `Agg` and `TargetAgg`! Tell me what you think about this.